### PR TITLE
[Feature] Support MultiValLoop

### PR DESCRIPTION
**Pull request by Czm369**
image

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/engine/runner/loops.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+from mmengine.model import is_model_wrapper
+from mmengine.runner import ValLoop
+
+from mmdet.registry import LOOPS
+
+
+@LOOPS.register_module()
+class MultiValLoop(ValLoop):
```
Comment (RangiLyu) [inline mmdet/engine/runner/loops.py:10]: This Loop requires the model has teacher and student. So it is not general enough to use this name. Maybe rename to `TeacherStudentValLoop`.

File: mmdet/engine/runner/loops.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    """Multi-loop for validation.
+
+    Args:
+        runner (Runner): A reference of runner.
+        dataloader (Dataloader or dict): A dataloader object or a dict to
+            build a dataloader.
+        evaluator (Evaluator or dict or list): Used for computing metrics.
+        fp16 (bool): Whether to enable fp16 validation. Defaults to
+            False.
```
Comment (RangiLyu) [inline mmdet/engine/runner/loops.py:19]: delete

File: mmdet/engine/runner/loops.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            for idx, data_batch in enumerate(self.dataloader):
+                self.run_iter(idx, data_batch)
+            # compute metrics
+            metrics = self.evaluator.evaluate(len(self.dataloader.dataset))
+            multi_metrics.update(
+                {'/'.join((_predict_on, k)): v
+                 for k, v in metrics.items()})
+        model.semi_test_cfg['predict_on'] = predict_on
+
+        self.runner.call_hook('after_val_epoch', metrics=multi_metrics)
```
Comment (hhaAndroid) [inline mmdet/engine/runner/loops.py:38]: Is it normal to log and store ckpt?

Comment (Czm369) [reply mmdet/engine/runner/loops.py:38]: It is normal to log and store ckpt, because `_load_from_state_dict` of semi-supervised models will be rewritten.

**Review by hhaAndroid — APPROVED** [review]

**Pull request merged**